pkg/internal/authorizer/authorizer.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2019 The KubeCarrier Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorizer
+
+import (
+	"context"
+	"fmt"
+
+	authv1 "k8s.io/api/authorization/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type RequestOperation string
+
+const (
+	RequestGet  RequestOperation = "get"
+	RequestList RequestOperation = "list"
+)
+
+type AuthorizationDecision string
+
+const (
+	DecisionAllowed   AuthorizationDecision = "allowed"
+	DecisionDenied    AuthorizationDecision = "denied"
+	DecisionNoOpinion AuthorizationDecision = "no opinion"
+)
+
+type Authorizer struct {
+	client.Client
+}
+
+type AuthorizationOption interface {
+	apply() *authv1.ResourceAttributes
+}
+
+type GetAuthorization struct {
+	Namespace string
+	Group     string
+	Version   string
+	Resource  string
+	Name      string
+}
+
+func (a GetAuthorization) apply() *authv1.ResourceAttributes {
+	attributes := &authv1.ResourceAttributes{}
+	attributes.Verb = string(RequestGet)
+	attributes.Namespace = a.Namespace
+	attributes.Group = a.Group
+	attributes.Version = a.Version
+	attributes.Resource = a.Resource
+	attributes.Name = a.Name
+	return attributes
+}
+
+type ListAuthorization struct {
+	Namespace string
+	Group     string
+	Version   string
+	Resource  string
+}
+
+func (a ListAuthorization) apply() *authv1.ResourceAttributes {
+	attributes := &authv1.ResourceAttributes{}
+	attributes.Verb = string(RequestList)
+	attributes.Namespace = a.Namespace
+	attributes.Group = a.Group
+	attributes.Version = a.Version
+	attributes.Resource = a.Resource
+	return attributes
+}
+
+func (a Authorizer) Authorize(
+	ctx context.Context,
+	user UserInfo,
+	option AuthorizationOption,
+) (AuthorizationDecision, error) {
+	review := &authv1.SubjectAccessReview{
+		Spec: authv1.SubjectAccessReviewSpec{
+			ResourceAttributes: option.apply(),
+			User:               user.Name,
+		},
+	}
+	if err := a.Create(ctx, review); err != nil {
+		return DecisionDenied, fmt.Errorf("creating SubjectAccessReview: %s", err)
+	}
+	switch {
+	case review.Status.Denied:
+		return DecisionDenied, nil
+	case review.Status.Allowed:
+		return DecisionAllowed, nil
+	default:
```

Comment:
When is this possible? The default should be Denied for security purposes....what the client should do if there's no opinion?

Reply:
the client only start operations if the Decision is DecisionAllowed, the `NoOpinion` is used for providing more detailed info.